### PR TITLE
SCLOPF for pyomo=False and ease of import

### LINKED ISSUE #173: Unable to import pypsa from v0.16.0 to v0.17.0
#169  Checklist

- The error happen from releases 0.16.0 to 0.17.0. The release 0.15.0 is working fine.
- Python: 2.7.17
- OS: Ubuntu 18.04 over WSL
- Installation from pip

## Describe the Bug

When trying to import pypsa errors are returned:

## Error Message

Error for release 0.16.0:
```
$ python2 comp.py
Traceback (most recent call last):
  File "comp.py", line 19, in <module>
    import pypsa
  File "/home/leonardo/.local/lib/python2.7/site-packages/pypsa/__init__.py", line 29, in <module>
    from . import pf, opf, opt, plot, networkclustering, io, contingency, geo, stats
  File "/home/leonardo/.local/lib/python2.7/site-packages/pypsa/stats.py", line 29
    mc[c] = c.df @ c.pnl['p']
                 ^
SyntaxError: invalid syntax
```

Error for releases 0.16.1 and 0.17.0:
```
$ python2 comp.py
Traceback (most recent call last):
  File "comp.py", line 19, in <module>
    import pypsa
  File "/home/leonardo/.local/lib/python2.7/site-packages/pypsa/__init__.py", line 25, in <module>
    from . import components, descriptors
  File "/home/leonardo/.local/lib/python2.7/site-packages/pypsa/components.py", line 45, in <module>
    from .io import (export_to_csv_folder, import_from_csv_folder,
  File "/home/leonardo/.local/lib/python2.7/site-packages/pypsa/io.py", line 749
    logger.warning(f"Components {diff} for attribute {attr} of {cls_name} "
                                                                          ^
SyntaxError: invalid syntax
```

### LINKED ISSUE #176: Security-constrained LOPF with pyomo=False
Security-constrained LOPF with `pyomo=False` should be fairly straightforward to implement. Just requires an nomopyomo variant of https://github.com/PyPSA/PyPSA/blob/master/pypsa/contingency.py#L230-L264 @shokriga

### PR DESCRIPTION
**Pull request by fneum**
Closes #176

## Changes proposed in this Pull Request

Re-implementation of security-constrained LOPF for `pyomo=False`.
- https://pypsa.readthedocs.io/en/latest/contingency_analysis.html#security-constrained-linear-optimal-power-flow-sclopf

Additionally, slight restructuring of `pypsa.contingency` to allow imports like

```python
from pypsa.contingency import add_contingency_constraints
from pypsa.contingency import add_contingency_constraints_lowmem
```
This would for instance simplify adding an SCLOPF option in PyPSA-Eur; just import and add to the collection of extra functionalities.

@FabianHofmann, I hope I did not misuse the `pypsa.linopt` functions. Particularly, I would be interested whether there would be a cleaner way to set the constraint reference.

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- ~[ ] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).~
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.

### CODE REVIEW

**Comment by fneum** [comment]
FYI @shokriga

**Comment by FabianHofmann** [comment]
On the first glimpse it looks good, I'll go through the code and test it tomorrow

File: pypsa/contingency.py [hunk 1]
```diff
@@ -154,11 +155,152 @@ def network_lpf_contingency(network, snapshots=None, branch_outages=None):
...
+        
+        BODF = pd.DataFrame(sn.BODF, index=branches_i, columns=branches_i)
+
+        lhs = {}
+        rhs = {}
+
+        for outage in outages:
+
+            def contingency_flow(branch):
+                if branch.name == outage: return ""
```
Comment (FabianHofmann) [inline pypsa/contingency.py:252]:
```suggestion
                if branch.name == outage:
                    return pd.Series('', branch.index)
```
I get some malformed pd.Series out of this, when this is applied to a pd.DataFrame (as below). I'm on pandas version 1.0.1, I think the `apply` API change during the the last big releases

File: pypsa/contingency.py [hunk 2]
```diff
@@ -154,11 +155,152 @@ def network_lpf_contingency(network, snapshots=None, branch_outages=None):
...
+                flow = linexpr((1, branch))
+                added_flow = linexpr((BODF.loc[branch.name, outage], p.loc[:, outage]))
```
Comment (FabianHofmann) [inline pypsa/contingency.py:254]:
```suggestion
                added_flow = linexpr((BODF.at[branch.name, outage], p.loc[:, outage]))
```
`at` is faster for single element indexing

**Comment by fneum** [comment]
About the style: absolutely right!

> I always use the define_constraints function which does not require an extra set_conref (which I find a bit annoying). In this case it seems a bit problematic as define_constraints would require a loop over the components which are stacked together.

Yes, my inspiration here was `define_kirchhoff_constraints` as it also loops over subnetworks to collect all constraints of the group.

**Comment by FabianHofmann** [comment]
> Yes, my inspiration here was `define_kirchhoff_constraints` as it also loops over subnetworks to collect all constraints of the group.

True, perhaps we can (at some point in the future) make `define_constraints`/`define_variables` robust for multiple components at one time, i.e. passing a list as name argument which all have to be included in the first level of the MultiIndex.

You have an idea why the test on windows machines fail? Local test work out fine.

**Comment by fneum** [comment]
> You have an idea why the test on windows machines fail? Local test work out fine.

My suspicion is that GLPK does not solve. I'll have a look (at least with GLPK on linux).

**Comment by fneum** [comment]
- [ ] Consider outage of only one parallel line. This is also not implemented in the `pyomo=True` version.

**Comment by fneum** [comment]
Problem with GLPK is now solved. Issue was that GLPK read-in requires at least one variable in LHS (other than CBC which can handle this case).

**Pull request merged**